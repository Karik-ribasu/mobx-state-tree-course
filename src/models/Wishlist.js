import { types, getParent, destroy } from "mobx-state-tree"

export const WishlistItem = types.model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, ""),
    id: types.number
})
    .actions(self => ({
        changeName(newName) {
            self.name = newName
        },
        changePrice(newPrice) {
            self.price = newPrice
        },
        changeImage(newImage) {
            self.image = newImage
        },
        remove() {
            getParent(self, 2).remove(self)
        }
    }))

export const Wishlist = types.model({
    items: types.optional(types.array(WishlistItem), []),
})
    .actions(self => ({
        add(newItem) {
            self.items.push(newItem)
        },
        remove(item) {
            console.log(item)
            destroy(item)
        }
    }))
    .views(self => ({
        get totalPrice() {
            return self.items.reduce((sum, entry) => sum + entry.price, 0)
        }
    }))