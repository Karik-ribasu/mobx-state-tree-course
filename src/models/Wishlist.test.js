import { reaction } from "mobx"
import { getSnapshot, onSnapshot } from "mobx-state-tree"
import { WishlistItem, Wishlist } from "./Wishlist"

it("Can create Wishlist Item", () => {

    const item = WishlistItem.create({
        name: "book",
        price: 50.99,
        id: 54896
    })
    expect(item.price).toBe(50.99)
    expect(item.name).toBe("book")
    item.changeName("table")
    expect(item.name).toBe("table")
})

it("Can create Wishlist", () => {
    const wishlist = Wishlist.create({
        items: [
            WishlistItem.create({
                name: "book",
                price: 50.99,
                image: "",
                id: 4864
            }),
            WishlistItem.create({
                name: "table",
                price: 210.99,
                image: "",
                id: 94315
            }),
        ]
    })

    expect(getSnapshot(wishlist)).toMatchSnapshot()
})

it("Can calculate total price", () => {
    const wishlist = Wishlist.create({
        items: [
            WishlistItem.create({
                name: "book",
                price: 50.99,
                image: ""
            }),
            WishlistItem.create({
                name: "table",
                price: 210.99,
                image: ""
            }),
        ]
    })

    let changed = 0
    reaction(() => wishlist.totalPrice, () => changed++)

    expect(wishlist.totalPrice).toBe(261.98)

    wishlist.items[0].changeName("pen")
    expect(changed).toBe(0)
    wishlist.items[0].changeImage("imagesample.jpg")
    expect(changed).toBe(0)
    wishlist.items[0].changePrice(2.99)
    expect(changed).toBe(1)
    console.log(wishlist.totalPrice)
})