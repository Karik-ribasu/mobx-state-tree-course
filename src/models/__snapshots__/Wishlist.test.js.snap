// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can create Wishlist 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "book",
      "price": 50.99,
    },
    Object {
      "image": "",
      "name": "table",
      "price": 210.99,
    },
  ],
}
`;
